finality-aleph/src/validator_network/protocols/v1/mod.rs

```diff
@@ -0,0 +1,446 @@
+    mut data_from_user: mpsc::UnboundedReceiver<D>,
+) -> Result<(), ProtocolError> {
+    use Message::*;
+    loop {
+        let to_send = match timeout(HEARTBEAT_TIMEOUT, data_from_user.next()).await {
+            Ok(maybe_data) => match maybe_data {
+                Some(data) => Data(data),
+                // We have been closed by the parent service, all good.
+                None => return Ok(()),
+            },
+            _ => Heartbeat,
+        };
+        sender = send_data(sender, to_send).await?;
+    }
+}
+
+async fn receiving<D: Data, S: AsyncRead + Unpin + Send>(
+    mut stream: S,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    use Message::*;
+    loop {
+        let (old_stream, message) = timeout(
+            MAX_MISSED_HEARTBEATS * HEARTBEAT_TIMEOUT,
+            receive_data(stream),
+        )
+        .await
+        .map_err(|_| ProtocolError::CardiacArrest)??;
+        stream = old_stream;
+        match message {
+            Data(data) => data_for_user
+                .unbounded_send(data)
+                .map_err(|_| ProtocolError::NoUserConnection)?,
+            Heartbeat => (),
+        }
+    }
+}
+
+async fn manage_connection<D: Data, S: AsyncWrite + Unpin + Send, R: AsyncRead + Unpin + Send>(
+    sender: S,
+    receiver: R,
+    data_from_user: mpsc::UnboundedReceiver<D>,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    let sending = sending(sender, data_from_user);
+    let receiving = receiving(receiver, data_for_user);
+    tokio::select! {
+        result = receiving => result,
+        result = sending => result,
+    }
+}
+
+/// Performs the outgoing handshake, and then manages a connection sending and receiving data.
+/// Exits on parent request, or in case of broken or dead network connection.
+pub async fn outgoing<D: Data, S: Splittable>(
+    stream: S,
+    authority_pen: AuthorityPen,
+    peer_id: AuthorityId,
+    result_for_parent: mpsc::UnboundedSender<ResultForService<D>>,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    trace!(target: "validator-network", "Extending hand to {}.", peer_id);
+    let (sender, receiver) = v0_handshake_outgoing(stream, authority_pen, peer_id.clone()).await?;
+    info!(target: "validator-network", "Outgoing handshake with {} finished successfully.", peer_id);
+    let (data_for_network, data_from_user) = mpsc::unbounded();
+    result_for_parent
+        .unbounded_send((peer_id.clone(), Some(data_for_network), ConnectionType::New))
+        .map_err(|_| ProtocolError::NoParentConnection)?;
+    debug!(target: "validator-network", "Starting worker for communicating with {}.", peer_id);
+    manage_connection(sender, receiver, data_from_user, data_for_user).await
+}
+
+/// Performs the incoming handshake, and then manages a connection sending and receiving data.
+/// Exits on parent request, or in case of broken or dead network connection.
```

Comment:
probably I have some blackout - how do we wait for parent request?

Reply:
When the data sending channel is dropped, clarified comment.